File: react_native_app/src/services/DatabaseService.js
/**
 * Database Service for WMS Mobile App
 * Manages local SQLite database for offline functionality
 * Integrates with MySQL backend as per user preference
 */

import SQLite from 'react-native-sqlite-storage';

// Enable SQLite debugging in development
SQLite.DEBUG(true);
SQLite.enablePromise(true);

class DatabaseServiceClass {
  constructor() {
    this.database = null;
    this.databaseName = 'WMSMobile.db';
    this.databaseVersion = '1.0';
    this.databaseDisplayName = 'WMS Mobile Database';
    this.databaseSize = 200000;
  }

  async initializeDatabase() {
    if (this.database) {
      console.log('Database already initialized');
      return this.database;
    }

    try {
      this.database = await SQLite.openDatabase({
        name: this.databaseName,
        location: 'default',
      });

      console.log('Database opened successfully');
      await this.createTables();
      return this.database;
    } catch (error) {
      console.error('Failed to open database:', error);
      throw error;
    }
  }

  async createTables() {
    const tables = [
      // Users table
      `CREATE TABLE IF NOT EXISTS users (
        id INTEGER PRIMARY KEY,
        username TEXT UNIQUE NOT NULL,
        email TEXT,
        first_name TEXT,
        last_name TEXT,
        role TEXT,
        branch_code TEXT,
        is_active INTEGER DEFAULT 1,
        created_at TEXT,
        updated_at TEXT,
        synced INTEGER DEFAULT 0
      )`,

      // GRPO Documents table
      `CREATE TABLE IF NOT EXISTS grpo_documents (
        id INTEGER PRIMARY KEY,
        po_number TEXT NOT NULL,
        supplier_code TEXT,
        supplier_name TEXT,
        warehouse_code TEXT,
        user_id INTEGER,
        qc_approver_id INTEGER,
        qc_approved_at TEXT,
        qc_notes TEXT,
        status TEXT DEFAULT 'draft',
        po_total REAL,
        sap_document_number TEXT,
        notes TEXT,
        created_at TEXT,
        updated_at TEXT,
        synced INTEGER DEFAULT 0,
        FOREIGN KEY (user_id) REFERENCES users (id)
      )`,

      // GRPO Items table
      `CREATE TABLE IF NOT EXISTS grpo_items (
        id INTEGER PRIMARY KEY,
        grpo_document_id INTEGER,
        line_number INTEGER,
        item_code TEXT,
        item_name TEXT,
        received_quantity REAL,
        ordered_quantity REAL,
        unit_of_measure TEXT,
        batch_number TEXT,
        expiration_date TEXT,
        serial_number TEXT,
        warehouse_code TEXT,
        bin_location TEXT,
        created_at TEXT,
        updated_at TEXT,
        synced INTEGER DEFAULT 0,
        FOREIGN KEY (grpo_document_id) REFERENCES grpo_documents (id)
      )`,

      // Inventory Transfers table
      `CREATE TABLE IF NOT EXISTS inventory_transfers (
        id INTEGER PRIMARY KEY,
        transfer_request_number TEXT,
        sap_document_number TEXT,
        status TEXT DEFAULT 'draft',
        user_id INTEGER,
        qc_approver_id INTEGER,
        qc_approved_at TEXT,
        qc_notes TEXT,
        from_warehouse TEXT,
        to_warehouse TEXT,
        transfer_type TEXT,
        priority TEXT,
        reason_code TEXT,
        notes TEXT,
        created_at TEXT,
        updated_at TEXT,
        synced INTEGER DEFAULT 0,
        FOREIGN KEY (user_id) REFERENCES users (id)
      )`,

      // Inventory Transfer Items table
      `CREATE TABLE IF NOT EXISTS inventory_transfer_items (
        id INTEGER PRIMARY KEY,
        inventory_transfer_id INTEGER,
        line_number INTEGER,
        item_code TEXT,
        item_name TEXT,
        quantity REAL,
        unit_of_measure TEXT,
        batch_number TEXT,
        from_bin_location TEXT,
        to_bin_location TEXT,
        created_at TEXT,
        updated_at TEXT,
        synced INTEGER DEFAULT 0,
        FOREIGN KEY (inventory_transfer_id) REFERENCES inventory_transfers (id)
      )`,

      // Pick Lists table
      `CREATE TABLE IF NOT EXISTS pick_lists (
        id INTEGER PRIMARY KEY,
        sales_order_number TEXT,
        customer_code TEXT,
        customer_name TEXT,
        warehouse_code TEXT,
        user_id INTEGER,
        status TEXT DEFAULT 'draft',
        priority TEXT,
        due_date TEXT,
        notes TEXT,
        created_at TEXT,
        updated_at TEXT,
        synced INTEGER DEFAULT 0,
        FOREIGN KEY (user_id) REFERENCES users (id)
      )`,

      // Pick List Items table
      `CREATE TABLE IF NOT EXISTS pick_list_items (
        id INTEGER PRIMARY KEY,
        pick_list_id INTEGER,
        line_number INTEGER,
        item_code TEXT,
        item_name TEXT,
        ordered_quantity REAL,
        picked_quantity REAL,
        unit_of_measure TEXT,
        batch_number TEXT,
        bin_location TEXT,
        created_at TEXT,
        updated_at TEXT,
        synced INTEGER DEFAULT 0,
        FOREIGN KEY (pick_list_id) REFERENCES pick_lists (id)
      )`,

      // Sync Queue table for offline operations
      `CREATE TABLE IF NOT EXISTS sync_queue (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        table_name TEXT NOT NULL,
        record_id INTEGER NOT NULL,
        operation TEXT NOT NULL,
        data TEXT,
        created_at TEXT DEFAULT CURRENT_TIMESTAMP,
        retry_count INTEGER DEFAULT 0,
        last_error TEXT
      )`,
    ];

    try {
      for (const sql of tables) {
        await this.database.executeSql(sql);
      }
      console.log('All tables created successfully');
    } catch (error) {
      console.error('Error creating tables:', error);
      throw error;
    }
  }

  // Generic CRUD operations
  async insert(tableName, data) {
    const columns = Object.keys(data).join(', ');
    const placeholders = Object.keys(data).map(() => '?').join(', ');
    const values = Object.values(data);

    const sql = `INSERT INTO ${tableName} (${columns}) VALUES (${placeholders})`;
    
    try {
      const result = await this.database.executeSql(sql, values);
      return result[0].insertId;
    } catch (error) {
      console.error(`Error inserting into ${tableName}:`, error);
      throw error;
    }
  }

  async update(tableName, data, whereClause, whereParams = []) {
    const setClause = Object.keys(data)
      .map(key => `${key} = ?`)
      .join(', ');
    const values = [...Object.values(data), ...whereParams];

    const sql = `UPDATE ${tableName} SET ${setClause} WHERE ${whereClause}`;
    
    try {
      const result = await this.database.executeSql(sql, values);
      return result[0].rowsAffected;
    } catch (error) {
      console.error(`Error updating ${tableName}:`, error);
      throw error;
    }
  }

  async select(tableName, whereClause = '', params = [], orderBy = '') {
    let sql = `SELECT * FROM ${tableName}`;
    if (whereClause) sql += ` WHERE ${whereClause}`;
    if (orderBy) sql += ` ORDER BY ${orderBy}`;

    try {
      const result = await this.database.executeSql(sql, params);
      const rows = [];
      for (let i = 0; i < result[0].rows.length; i++) {
        rows.push(result[0].rows.item(i));
      }
      return rows;
    } catch (error) {
      console.error(`Error selecting from ${tableName}:`, error);
      throw error;
    }
  }

  async delete(tableName, whereClause, params = []) {
    const sql = `DELETE FROM ${tableName} WHERE ${whereClause}`;
    
    try {
      const result = await this.database.executeSql(sql, params);
      return result[0].rowsAffected;
    } catch (error) {
      console.error(`Error deleting from ${tableName}:`, error);
      throw error;
    }
  }

  // GRPO specific operations
  async getGRPODocuments(userId = null) {
    const whereClause = userId ? 'user_id = ?' : '';
    const params = userId ? [userId] : [];
    return await this.select('grpo_documents', whereClause, params, 'created_at DESC');
  }

  async createGRPODocument(grpoData) {
    const data = {
      ...grpoData,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString(),
      synced: 0,
    };
    return await this.insert('grpo_documents', data);
  }

  async updateGRPODocument(id, grpoData) {
    const data = {
      ...grpoData,
      updated_at: new Date().toISOString(),
      synced: 0,
    };
    return await this.update('grpo_documents', data, 'id = ?', [id]);
  }

  // Inventory Transfer specific operations
  async getInventoryTransfers(userId = null) {
    const whereClause = userId ? 'user_id = ?' : '';
    const params = userId ? [userId] : [];
    return await this.select('inventory_transfers', whereClause, params, 'created_at DESC');
  }

  async createInventoryTransfer(transferData) {
    const data = {
      ...transferData,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString(),
      synced: 0,
    };
    return await this.insert('inventory_transfers', data);
  }

  async updateInventoryTransfer(id, transferData) {
    const data = {
      ...transferData,
      updated_at: new Date().toISOString(),
      synced: 0,
    };
    return await this.update('inventory_transfers', data, 'id = ?', [id]);
  }

  // Pick List specific operations
  async getPickLists(userId = null) {
    const whereClause = userId ? 'user_id = ?' : '';
    const params = userId ? [userId] : [];
    return await this.select('pick_lists', whereClause, params, 'created_at DESC');
  }

  async createPickList(pickListData) {
    const data = {
      ...pickListData,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString(),
      synced: 0,
    };
    return await this.insert('pick_lists', data);
  }

  async updatePickList(id, pickListData) {
    const data = {
      ...pickListData,
      updated_at: new Date().toISOString(),
      synced: 0,
    };
    return await this.update('pick_lists', data, 'id = ?', [id]);
  }

  // Sync queue operations
  async addToSyncQueue(tableName, recordId, operation, data = null) {
    const queueData = {
      table_name: tableName,
      record_id: recordId,
      operation: operation,
      data: data ? JSON.stringify(data) : null,
      created_at: new Date().toISOString(),
    };
    return await this.insert('sync_queue', queueData);
  }

  async getSyncQueue() {
    return await this.select('sync_queue', '', [], 'created_at ASC');
  }

  async removeSyncQueueItem(id) {
    return await this.delete('sync_queue', 'id = ?', [id]);
  }

  async closeDatabase() {
    if (this.database) {
      await this.database.close();
      this.database = null;
      console.log('Database closed');
    }
  }
}

export const DatabaseService = new DatabaseServiceClass();

File: mobile_app/lib/services/database_service.dart
import 'dart:async';
import 'package:sqflite/sqflite.dart';
import 'package:path/path.dart';
import 'package:path_provider/path_provider.dart';
import '../models/user.dart';
import '../models/inventory_transfer.dart';
import '../models/grpo_document.dart';

class DatabaseService {
  static Database? _database;
  static const String _databaseName = 'wms_mobile.db';
  static const int _databaseVersion = 1;

  Future<Database> get database async {
    _database ??= await _initDatabase();
    return _database!;
  }

  Future<void> initDatabase() async {
    if (_database == null) {
      await database;
    }
  }

  Future<Database> _initDatabase() async {
    final documentsDirectory = await getApplicationDocumentsDirectory();
    final path = join(documentsDirectory.path, _databaseName);
    
    return await openDatabase(
      path,
      version: _databaseVersion,
      onCreate: _onCreate,
      onUpgrade: _onUpgrade,
    );
  }

  Future<void> _onCreate(Database db, int version) async {
    print('📱 Creating local database tables...');
    
    // Users table
    await db.execute('''
      CREATE TABLE users (
        id INTEGER PRIMARY KEY,
        username TEXT UNIQUE NOT NULL,
        email TEXT UNIQUE NOT NULL,
        first_name TEXT,
        last_name TEXT,
        role TEXT NOT NULL,
        branch_code TEXT,
        is_active INTEGER DEFAULT 1,
        created_at TEXT NOT NULL,
        updated_at TEXT NOT NULL,
        sync_status INTEGER DEFAULT 1,
        last_synced_at TEXT
      )
    ''');

    // Inventory Transfers table
    await db.execute('''
      CREATE TABLE inventory_transfers (
        id INTEGER PRIMARY KEY,
        transfer_request_number TEXT NOT NULL,
        sap_document_number TEXT,
        status TEXT NOT NULL,
        user_id INTEGER NOT NULL,
        qc_approver_id INTEGER,
        qc_approved_at TEXT,
        qc_notes TEXT,
        from_warehouse TEXT,
        to_warehouse TEXT,
        transfer_type TEXT NOT NULL,
        priority TEXT NOT NULL,
        reason_code TEXT,
        notes TEXT,
        created_at TEXT NOT NULL,
        updated_at TEXT NOT NULL,
        sync_status INTEGER DEFAULT 0,
        last_synced_at TEXT,
        FOREIGN KEY (user_id) REFERENCES users (id),
        FOREIGN KEY (qc_approver_id) REFERENCES users (id)
      )
    ''');

    // GRPO Documents table
    await db.execute('''
      CREATE TABLE grpo_documents (
        id INTEGER PRIMARY KEY,
        po_number TEXT NOT NULL,
        supplier_code TEXT,
        supplier_name TEXT,
        warehouse_code TEXT,
        user_id INTEGER NOT NULL,
        qc_approver_id INTEGER,
        qc_approved_at TEXT,
        qc_notes TEXT,
        status TEXT NOT NULL,
        po_total REAL,
        sap_document_number TEXT,
        notes TEXT,
        created_at TEXT NOT NULL,
        updated_at TEXT NOT NULL,
        sync_status INTEGER DEFAULT 0,
        last_synced_at TEXT,
        FOREIGN KEY (user_id) REFERENCES users (id),
        FOREIGN KEY (qc_approver_id) REFERENCES users (id)
      )
    ''');

    // Transfer Items table
    await db.execute('''
      CREATE TABLE transfer_items (
        id INTEGER PRIMARY KEY,
        transfer_id INTEGER NOT NULL,
        item_code TEXT NOT NULL,
        item_name TEXT,
        quantity REAL NOT NULL,
        unit_of_measure TEXT,
        from_warehouse_code TEXT,
        to_warehouse_code TEXT,
        from_bin TEXT,
        to_bin TEXT,
        batch_number TEXT,
        serial_number TEXT,
        expiry_date TEXT,
        unit_price REAL,
        total_value REAL,
        qc_status TEXT DEFAULT 'pending',
        base_entry INTEGER,
        base_line INTEGER,
        sap_line_number INTEGER,
        created_at TEXT NOT NULL,
        updated_at TEXT NOT NULL,
        sync_status INTEGER DEFAULT 0,
        FOREIGN KEY (transfer_id) REFERENCES inventory_transfers (id) ON DELETE CASCADE
      )
    ''');

    // GRPO Items table
    await db.execute('''
      CREATE TABLE grpo_items (
        id INTEGER PRIMARY KEY,
        grpo_id INTEGER NOT NULL,
        item_code TEXT NOT NULL,
        item_name TEXT,
        quantity REAL NOT NULL,
        received_quantity REAL DEFAULT 0,
        unit_price REAL,
        line_total REAL,
        unit_of_measure TEXT,
        warehouse_code TEXT,
        bin_location TEXT,
        batch_number TEXT,
        serial_number TEXT,
        expiry_date TEXT,
        barcode TEXT,
        qc_status TEXT DEFAULT 'pending',
        po_line_number INTEGER,
        base_entry INTEGER,
        base_line INTEGER,
        created_at TEXT NOT NULL,
        updated_at TEXT NOT NULL,
        sync_status INTEGER DEFAULT 0,
        FOREIGN KEY (grpo_id) REFERENCES grpo_documents (id) ON DELETE CASCADE
      )
    ''');

    // Sync Queue table for offline operations
    await db.execute('''
      CREATE TABLE sync_queue (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        table_name TEXT NOT NULL,
        record_id INTEGER NOT NULL,
        operation TEXT NOT NULL, -- 'CREATE', 'UPDATE', 'DELETE'
        data TEXT, -- JSON data
        created_at TEXT NOT NULL,
        retry_count INTEGER DEFAULT 0,
        last_error TEXT
      )
    ''');

    print('✅ Database tables created successfully');
  }

  Future<void> _onUpgrade(Database db, int oldVersion, int newVersion) async {
    print('📱 Upgrading database from version $oldVersion to $newVersion');
    // Handle database upgrades here
  }

  // User operations
  Future<int> insertUser(User user) async {
    final db = await database;
    return await db.insert(
      'users',
      _userToMap(user),
      conflictAlgorithm: ConflictAlgorithm.replace,
    );
  }

  Future<User?> getUser(int id) async {
    final db = await database;
    final List<Map<String, dynamic>> maps = await db.query(
      'users',
      where: 'id = ?',
      whereArgs: [id],
    );

    if (maps.isNotEmpty) {
      return _userFromMap(maps.first);
    }
    return null;
  }

  Future<List<User>> getAllUsers() async {
    final db = await database;
    final List<Map<String, dynamic>> maps = await db.query('users');
    return List.generate(maps.length, (i) => _userFromMap(maps[i]));
  }

  // Inventory Transfer operations
  Future<int> insertInventoryTransfer(InventoryTransfer transfer) async {
    final db = await database;
    return await db.insert(
      'inventory_transfers',
      _transferToMap(transfer),
      conflictAlgorithm: ConflictAlgorithm.replace,
    );
  }

  Future<List<InventoryTransfer>> getAllInventoryTransfers() async {
    final db = await database;
    final List<Map<String, dynamic>> maps = await db.query(
      'inventory_transfers',
      orderBy: 'created_at DESC',
    );
    return List.generate(maps.length, (i) => _transferFromMap(maps[i]));
  }

  Future<InventoryTransfer?> getInventoryTransfer(int id) async {
    final db = await database;
    final List<Map<String, dynamic>> maps = await db.query(
      'inventory_transfers',
      where: 'id = ?',
      whereArgs: [id],
    );

    if (maps.isNotEmpty) {
      return _transferFromMap(maps.first);
    }
    return null;
  }

  Future<List<InventoryTransfer>> getInventoryTransfersByUser(int userId) async {
    final db = await database;
    final List<Map<String, dynamic>> maps = await db.query(
      'inventory_transfers',
      where: 'user_id = ?',
      whereArgs: [userId],
      orderBy: 'created_at DESC',
    );
    return List.generate(maps.length, (i) => _transferFromMap(maps[i]));
  }

  Future<int> updateInventoryTransfer(InventoryTransfer transfer) async {
    final db = await database;
    return await db.update(
      'inventory_transfers',
      _transferToMap(transfer),
      where: 'id = ?',
      whereArgs: [transfer.id],
    );
  }

  Future<int> deleteInventoryTransfer(int id) async {
    final db = await database;
    return await db.delete(
      'inventory_transfers',
      where: 'id = ?',
      whereArgs: [id],
    );
  }

  // GRPO Document operations
  Future<int> insertGRPODocument(GRPODocument grpo) async {
    final db = await database;
    return await db.insert(
      'grpo_documents',
      _grpoToMap(grpo),
      conflictAlgorithm: ConflictAlgorithm.replace,
    );
  }

  Future<List<GRPODocument>> getAllGRPODocuments() async {
    final db = await database;
    final List<Map<String, dynamic>> maps = await db.query(
      'grpo_documents',
      orderBy: 'created_at DESC',
    );
    return List.generate(maps.length, (i) => _grpoFromMap(maps[i]));
  }

  Future<GRPODocument?> getGRPODocument(int id) async {
    final db = await database;
    final List<Map<String, dynamic>> maps = await db.query(
      'grpo_documents',
      where: 'id = ?',
      whereArgs: [id],
    );

    if (maps.isNotEmpty) {
      return _grpoFromMap(maps.first);
    }
    return null;
  }

  Future<List<GRPODocument>> getGRPODocumentsByUser(int userId) async {
    final db = await database;
    final List<Map<String, dynamic>> maps = await db.query(
      'grpo_documents',
      where: 'user_id = ?',
      whereArgs: [userId],
      orderBy: 'created_at DESC',
    );
    return List.generate(maps.length, (i) => _grpoFromMap(maps[i]));
  }

  // Sync operations
  Future<List<InventoryTransfer>> getPendingSyncTransfers() async {
    final db = await database;
    final List<Map<String, dynamic>> maps = await db.query(
      'inventory_transfers',
      where: 'sync_status = ?',
      whereArgs: [0], // 0 = pending sync
    );
    return List.generate(maps.length, (i) => _transferFromMap(maps[i]));
  }

  Future<void> markTransferAsSynced(int id) async {
    final db = await database;
    await db.update(
      'inventory_transfers',
      {
        'sync_status': 1, // 1 = synced
        'last_synced_at': DateTime.now().toIso8601String(),
      },
      where: 'id = ?',
      whereArgs: [id],
    );
  }

  Future<void> clearDatabase() async {
    final db = await database;
    await db.delete('users');
    await db.delete('inventory_transfers');
    await db.delete('grpo_documents');
    await db.delete('transfer_items');
    await db.delete('grpo_items');
    await db.delete('sync_queue');
  }

  // Helper methods for data conversion
  Map<String, dynamic> _userToMap(User user) {
    return {
      'id': user.id,
      'username': user.username,
      'email': user.email,
      'first_name': user.firstName,
      'last_name': user.lastName,
      'role': user.role,
      'branch_code': user.branchCode,
      'is_active': user.isActive ? 1 : 0,
      'created_at': user.createdAt.toIso8601String(),
      'updated_at': user.updatedAt.toIso8601String(),
      'sync_status': 1,
      'last_synced_at': DateTime.now().toIso8601String(),
    };
  }

  User _userFromMap(Map<String, dynamic> map) {
    return User(
      id: map['id'],
      username: map['username'],
      email: map['email'],
      firstName: map['first_name'],
      lastName: map['last_name'],
      role: map['role'],
      branchCode: map['branch_code'],
      isActive: map['is_active'] == 1,
      createdAt: DateTime.parse(map['created_at']),
      updatedAt: DateTime.parse(map['updated_at']),
    );
  }

  Map<String, dynamic> _transferToMap(InventoryTransfer transfer) {
    return {
      'id': transfer.id,
      'transfer_request_number': transfer.transferRequestNumber,
      'sap_document_number': transfer.sapDocumentNumber,
      'status': transfer.status,
      'user_id': transfer.userId,
      'qc_approver_id': transfer.qcApproverId,
      'qc_approved_at': transfer.qcApprovedAt?.toIso8601String(),
      'qc_notes': transfer.qcNotes,
      'from_warehouse': transfer.fromWarehouse,
      'to_warehouse': transfer.toWarehouse,
      'transfer_type': transfer.transferType,
      'priority': transfer.priority,
      'reason_code': transfer.reasonCode,
      'notes': transfer.notes,
      'created_at': transfer.createdAt.toIso8601String(),
      'updated_at': transfer.updatedAt.toIso8601String(),
      'sync_status': transfer.syncStatus,
      'last_synced_at': transfer.lastSyncedAt?.toIso8601String(),
    };
  }

  InventoryTransfer _transferFromMap(Map<String, dynamic> map) {
    return InventoryTransfer(
      id: map['id'],
      transferRequestNumber: map['transfer_request_number'],
      sapDocumentNumber: map['sap_document_number'],
      status: map['status'],
      userId: map['user_id'],
      qcApproverId: map['qc_approver_id'],
      qcApprovedAt: map['qc_approved_at'] != null 
          ? DateTime.parse(map['qc_approved_at']) 
          : null,
      qcNotes: map['qc_notes'],
      fromWarehouse: map['from_warehouse'],
      toWarehouse: map['to_warehouse'],
      transferType: map['transfer_type'],
      priority: map['priority'],
      reasonCode: map['reason_code'],
      notes: map['notes'],
      createdAt: DateTime.parse(map['created_at']),
      updatedAt: DateTime.parse(map['updated_at']),
      syncStatus: map['sync_status'] ?? 0,
      lastSyncedAt: map['last_synced_at'] != null 
          ? DateTime.parse(map['last_synced_at']) 
          : null,
    );
  }

  Map<String, dynamic> _grpoToMap(GRPODocument grpo) {
    return {
      'id': grpo.id,
      'po_number': grpo.poNumber,
      'supplier_code': grpo.supplierCode,
      'supplier_name': grpo.supplierName,
      'warehouse_code': grpo.warehouseCode,
      'user_id': grpo.userId,
      'qc_approver_id': grpo.qcApproverId,
      'qc_approved_at': grpo.qcApprovedAt?.toIso8601String(),
      'qc_notes': grpo.qcNotes,
      'status': grpo.status,
      'po_total': grpo.poTotal,
      'sap_document_number': grpo.sapDocumentNumber,
      'notes': grpo.notes,
      'created_at': grpo.createdAt.toIso8601String(),
      'updated_at': grpo.updatedAt.toIso8601String(),
      'sync_status': grpo.syncStatus,
      'last_synced_at': grpo.lastSyncedAt?.toIso8601String(),
    };
  }

  GRPODocument _grpoFromMap(Map<String, dynamic> map) {
    return GRPODocument(
      id: map['id'],
      poNumber: map['po_number'],
      supplierCode: map['supplier_code'],
      supplierName: map['supplier_name'],
      warehouseCode: map['warehouse_code'],
      userId: map['user_id'],
      qcApproverId: map['qc_approver_id'],
      qcApprovedAt: map['qc_approved_at'] != null 
          ? DateTime.parse(map['qc_approved_at']) 
          : null,
      qcNotes: map['qc_notes'],
      status: map['status'],
      poTotal: map['po_total'],
      sapDocumentNumber: map['sap_document_number'],
      notes: map['notes'],
      createdAt: DateTime.parse(map['created_at']),
      updatedAt: DateTime.parse(map['updated_at']),
      syncStatus: map['sync_status'] ?? 0,
      lastSyncedAt: map['last_synced_at'] != null 
          ? DateTime.parse(map['last_synced_at']) 
          : null,
    );
  }
}